peps/pep-806.rst

```diff
@@ -0,0 +1,292 @@
+switch between sync and async context managers.  At the same time, it uses
+only existing keywords, distinguishing async code with the ``async`` keyword
+more precisely even than our current syntax.
+
+We do not propose that the ``async with`` statement should ever be deprecated,
+and indeed advocate its continued use for single-line statements so that
+"async" is the first non-whitespace token of each line opening an async
+context manager.
+
+
+Real-World Impact
+-----------------
+
+These enhancements address pain points that Python developers encounter daily.
+We surveyed an industry codebase, finding more than ten thousand functions
+containing at least one async context manager.  19% of these also contained a
+sync context manager.  For reference, async functions contain sync context
+managers about two-thirds as often as they contain async context managers.
+
+39% of functions with both ``with`` and ``async with`` statements could switch
+immediately to the proposed syntax, but this is a loose lower
+bound due to avoidance of sync context managers and use of workarounds listed
+under Rejected Ideas.  Based on inspecting a random sample of functions, we
+estimate that between 20% and 50% of async functions containing any context
+manager would use ``with async`` if this PEP is accepted.
+
+Across the ecosystem more broadly, we expect lower rates, perhaps in the
+5% to 20% range: the surveyed codebase uses structured concurrency with Trio,
+and also makes extensive use of context managers to mitigate the issues
+discussed in :pep:`533` and :pep:`789`.
+
+
+Rationale
+=========
+
+Mixed sync/async context managers are common in modern Python applications,
+such as async database connections or API clients and synchronous file
+operations.  The current syntax forces developers to choose between deeply
+nested code or error-prone workarounds like :class:`~contextlib.AsyncExitStack`.
+
+This PEP addresses the problem with a minimal syntax change that builds on
+existing patterns. By allowing individual context managers to be marked with
+``async``, we maintain Python's explicit approach to asynchronous code while
+eliminating unnecessary nesting.
+
+The implementation as syntactic sugar ensures zero runtime overhead - the new
+syntax desugars to the same nested ``with`` and ``async with`` statements
+developers write today. This approach requires no new protocols, no changes
+to existing context managers, and no new runtime behaviors to understand.
+
+
+Specification
+=============
+
+The ``with ..., async ...:`` syntax desugars into a sequence of context
+managers in the same way as current multi-context ``with`` statements,
+except that those prefixed by the ``async`` keyword use the ``__aenter__`` /
+``__aexit__`` protocol.
+
+Only the ``with`` statement is modified; ``async with async ctx():`` is a
+syntax error.
+
+The :class:`ast.withitem` node gains a new ``is_async`` integer attribute,
```

Comment:
Why not `bool`?

Reply:
I felt that [consistency with `ast.comprehension.is_async`](https://docs.python.org/3/library/ast.html#ast.comprehension) outweighed my preference for `bool` - while I'd support refactoring the `ast` module to use `is_async: bool` everywhere, I feel that's a separate question from this PEP.  I can add a footnote about it though.

Reply:
I don't think the AST meta-language supports booleans.